core/sql/optimizer/BindRelExpr.cpp

```diff
@@ -10272,8 +10273,14 @@ RelExpr* Insert::xformUpsertToMerge(BindWA *bindWA)
   ((Join*)re)->doNotTransformToTSJ();
   ((Join*)re)->setTSJForMerge(TRUE);	
   ((Join*)re)->setTSJForMergeWithInsert(TRUE);
+  ((Join*)re)->setTSJForMergeUpsert(TRUE);
   ((Join*)re)->setTSJForWrite(TRUE);
-  if (bindWA->hasDynamicRowsetsInQuery())
+
+  // if Inputs of current insert are empty (i.e. we have no params/rowsets)
+  // then there will be no pull up of inputs during transform and the join will
+  // not see the inputs of the mergeUpdate due to intermediate nodes. So
+  // add inputs directly to join and use elimination to remove extra inputs
+  if (NOT getGroupAttr()->getCharacteristicInputs().isEmpty())
```

Comment:
I'm having trouble understanding this if statement. As far as I can see, _myOuterRefs_ are the source values, coming from the child of the original insert. After the transformation they will come from the left child of the tuple flow, _re_, and be passed to the right child, _mu_. So, I would have added them unconditionally as char. inputs of _mu_. If the insert (_this_) char. inputs are non-empty, then I would have add those to _re_ and maybe even to _mu_.